oscilloscope/agilent.py

```diff
@@ -0,0 +1,195 @@
+        
+    # TRIGGER
+
+    TRIGGER_SOURCES = ['EXT', 'LINE', 'CHAN1', 'CHAN2', 'CHAN3', 'CHAN4']
+    trigger_source_write = 'TRIG:EDGE:SOUR {0}'.format
+    trigger_source_query = 'TRIG:EDGE:SOUR?'.format
+    @staticmethod
+    def trigger_source_parse(x):
+        return x
+
+    TRIGGER_SLOPES = ['POS', 'NEG']
+    trigger_slope_write = ':TRIG:EDGE:SLOP {}'.format
+    trigger_slope_query = ':TRIG:EDGE:SLOP?'.format
+    @staticmethod
+    def trigger_slope_parse(x):
+        return x
+
+    TRIGGER_LEVEL_UNIT = U.Unit('V')
+    @classmethod
+    def trigger_level_write(cls, level):
+        return ':TRIG:EDGE:LEV {}'.format(level[cls.TRIGGER_LEVEL_UNIT])
+    @staticmethod
+    def trigger_level_query():
+        return ':TRIG:EDGE:LEV?'
+    @classmethod
+    def trigger_level_parse(cls, level):
+        return float(level) * cls.TRIGGER_LEVEL_UNIT
+
+    @inlineCallbacks
+    def get_trace(self, channel):
+        """Get a trace.
+
+        Args:
+            channel (int): Which channel to get.
+
+        Returns:
+            (ValueArray[s]): Time axis.
+            (ValueArray[V]): Voltage axis.
+        """
+        word_length = 2
+        yield self.write('WAV:SOUR CHAN{}'.format(channel))
+        yield self.write('WAV:MSBF')
+        preamble_raw = yield self.query(':WAV:PRE?')
+        preamble = parse_preamble(preamble_raw)
+        yield self.write('WAV:FORM WORD')
+        yield self.write('WAV:DATA?')
+        wave_bin = yield self.read_raw()
+        wave_raw = parse_binary_waveform(wave_bin, word_length=word_length)
+
+        y_incriment = preamble['y_incriment']
+        y_origin = preamble['y_origin']
+        y_reference = preamble['y_reference']
+        wave_volts = ((wave_raw - y_reference) * y_incriment) + y_origin
+
+        x_incriment = preamble['x_incriment']
+        x_origin = preamble['x_origin']
+        num_points = preamble['num_points']
+        time_s = np.linspace(
+                x_origin,
+                x_origin + (num_points-1) * x_incriment,
+                num_points)
+        returnValue((time_s * U.Unit('s'), wave_volts * U.Unit('V')))
+ 
+
+def parse_preamble(preamble):
+    """Parse the waveform preamble.
+
+    Args:
+        preamble (str): ASCII encoded preamble.
+
+    Returns:
+        (dict): Mapping field names to their values.
+    """
+    fields = preamble.split(',')
+    num_points = int(fields[2])
+    x_incriment = float(fields[4])
+    x_origin = float(fields[5])
+    x_reference = int(fields[6])
+    y_incriment = float(fields[7])
+    y_origin = float(fields[8])
+    y_reference = int(fields[9])
+    return {'num_points': num_points,
+            'x_incriment': x_incriment,
+            'x_origin': x_origin,
+            'x_reference': x_reference,
+            'y_incriment': y_incriment,
+            'y_origin': y_origin,
+            'y_reference': y_reference}
+
+
+def parse_binary_waveform(data, word_length):
+    """Convert binary waveform dump to numpy array.
```

Comment:
As written, this parse binary waveform function only works when :WAVform:STReaming is off.  We should probably support both.  [Pages 1091-1092 in the Infiniium manual describe both binary block formats](http://www.keysight.com/upload/cmc_upload/All/9000_series_prog_ref.pdf?&cc=US&lc=eng#page=1091).